src/VisualStudio/Core/Def/Implementation/Diagnostics/FxCopAnalyzersSuggestedActionCallback.cs

```diff
@@ -0,0 +1,282 @@
+            if (!_infoBarChecked &&
+                !IsVsixInstalled() &&
+                !IsNuGetInstalled() &&
+                IsCandidate())
+            {
+                ShowInfoBarIfNecessary();
+            }
+        }
+
+        private bool IsVsixInstalled()
+        {
+            if (_vsixInstallStatus == FxCopAnalyzersInstallStatus.Unknown)
+            {
+                var vsShell = _serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+                var hr = vsShell.IsPackageInstalled(FxCopAnalyzersPackageGuid, out var installed);
+                if (ErrorHandler.Failed(hr))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+
+                    // We set installed to ensure we don't go through this again next time a
+                    // suggested action is called, and we don't want to continue if the shell
+                    // is busted.
+                    _vsixInstallStatus = FxCopAnalyzersInstallStatus.Installed;
+                }
+                else
+                {
+                    _vsixInstallStatus = installed != 0 ? FxCopAnalyzersInstallStatus.Installed : FxCopAnalyzersInstallStatus.NotInstalled;
+                    FxCopAnalyzersInstallLogger.LogVsixInstallationStatus(_workspace, _vsixInstallStatus);
+                }
+            }
+
+            return _vsixInstallStatus == FxCopAnalyzersInstallStatus.Installed;
+        }
+
+        private bool IsNuGetInstalled()
+        {
+            if (_nugetInstallStatusForCurrentSolution != FxCopAnalyzersInstallStatus.Installed)
+            {
+                var activeDocumentId = _documentTrackingService.TryGetActiveDocument();
+                if (activeDocumentId == null)
+                {
+                    return false;
+                }
+
+                var document = _workspace.CurrentSolution.GetTextDocument(activeDocumentId);
+                if (document == null)
+                {
+                    return false;
+                }
+
+                foreach (var analyzerReference in document.Project.AnalyzerReferences)
+                {
+                    if (ChildAnalyzerNuGetPackageIds.Contains(analyzerReference.Display))
+                    {
+                        // We set installed to ensure we don't go through this again next time a
+                        // suggested action is called for any document in current solution.
+                        _nugetInstallStatusForCurrentSolution = FxCopAnalyzersInstallStatus.Installed;
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private bool IsCandidate()
+        {
+            // Candidates fill the following critera:
+            //     1: Are a Dotnet user (as evidenced by the fact that this code is being run)
+            //     2: Have triggered a lightbulb on 3 separate days
```

Comment:
> Have triggered a lightbulb on 3 separate days

@jinujoseph @vatsalyaagrawal Do we want to make this logic more aggressive?

Reply:
I'd recommend keeping things non-aggressive :)  aggressive software feels bad.  You'll certainly get tons of eyeballs with the current approach. :)

Reply:
@CyrusNajmabadi Thanks. What I meant by aggressive was tuning the logic when the info bar shows up. For example, we have recently added bunch of code quality rules in the IDE, so when a user invokes a code fix for one of these diagnostics, it is more likely they want to enable additional code quality analyses if presented an option at that time, rather then a random light bulb invocation few sessions and days down the line. So we probably should not wait for multiple different VS sessions over multiple days before prompting them about additional available analyses. I am not proposing we change the existing non-aggressive design of ensuring we do not show the info bar more then once a day and also respect the Never Show Again option from current and past releases

Reply:
Couple more things to note:

1. Ever since the effort to create FxCop analyzers, the management has had thoughts of enabling the FxCop analyzers VSIX in Visual Studio IDE by default, which is way more aggressive then any of the current approaches. It was always thought to be complimentary with the existing code style rules in the IDE, and something that seemed valuable to include in the box. There are couple of things preventing us from doing so:
   1. RPS image loads: FxCop analyzers is spread across multiple language agnostic and language dependent assemblies, so this would incur multiple assembly loads.
   2. Customer stance: We want to enable these additional analyses by default only if we know for sure that customers are fine with additional suggestions showing up in their messages tab in error list to opportunistically improve code quality. @kendrahavens and @mikadumont are going to create a customer survey to find out how the customers feel about additional analyses/suggestions being added to the IDE with every release: useful. noise, don't care?

2. This PR is based on the strong positive feedback we received for FxCop Analyzers VSIX on top of VS2017. The current info bar was the only discoverability point for this VSIX, and when it stopped working in VS2019, we got strong community requests to support the VSIX on VS2019 (see https://github.com/dotnet/roslyn-analyzers/issues/2034, and their were many other duplicates of this issue). We also received extremely large number of bugs on the analyzers repo, with the repro steps citing the VSIX as the analyzer source. Additionally, we haven't received any VSFeedback about this info bar being annoying or aggressive. On the contrary, there have been extremely large number of feedback requests internally and externally on Microsoft's recommended code quality analyses (FxCop analyzers) being undiscoverable and not easy to search/install within Visual Studio. We are aiming to fill these discoverability gaps in 16.3.